donkeycar/parts/gps.py

```diff
@@ -2,163 +2,190 @@
-        returns line if read and None if no line was read
-        """
-        if self.lock.acquire(blocking=False):
-            try:
-                # TODO: Serial.in_waiting _always_ returns 0 in Macintosh
-                if self.gps is not None and self.gps.is_open and (is_mac() or self.gps.in_waiting):
-                    return self.gps.readline().decode()
-            except serial.serialutil.SerialException:
-                pass
-            except UnicodeDecodeError:
-                # the first sentence often includes mis-framed garbase
-                pass  # ignore and keep going
-            finally:
-                self.lock.release()
-        return None
+    def run_threaded(self, lines):
+        return self.run(lines)
 
-    def poll(self, timestamp=None):
-        #
-        # read a line and convert to a position
-        # in a threadsafe manner
-        #
-        # if there are characters waiting
-        # then read the line and parse it
-        #
-        if self.running:
-            if timestamp is None:
-                timestamp = time.time()
-            line = self._readline()
-            if line:
-                if self.debug:
-                    logger.info(line)
-                position = getGpsPosition(line, debug=self.debug)
-                if position:
-                    # (timestamp, longitude latitude)
-                    return (timestamp, position[0], position[1])
-        return None
+class GpsLatestPosition:
+    """
+    Return most recent valid GPS position
+    """
+    def __init__(self, debug=False):
+        self.debug = debug
+        self.position = None
 
-    def run(self):
-        if self.running:
-            #
-            # in non-threaded mode, just read a single reading and return it
-            #
-            if self.gps is not None:
-                position = self.poll(time.time())
-                if position:
-                    # [(timestamp, longitude, latitude)]
-                    return [position]
+    def run(self, positions):
+        if positions is not None and len(positions) > 0:
+            self.position = positions[-1]
+        return self.position
 
-        return []
+class GpsPosition:
+    """
+    Donkeycar part to read NMEA lines from serial port and convert a position
+    """
+    def __init__(self, serial:SerialPort, debug = False) -> None:
+        self.line_reader = SerialLineReader(serial)
+        self.debug = debug
+        self.position_reader = GpsNmeaPositions()
+        self.position = None
+        self._start()
+
+    def _start(self):
+        # wait until we get at least one gps position
+        while self.position is None:
+            print("Waiting for gps fix")
+            self.position = self.run()
+
+    def run_once(self, lines):
+        positions = self.GpsNmeaPositions.run(lines)
+        if positions is not None and len(positions) > 0:
+            self.position = positions[-1]
+            if self.debug:
```

Comment:
I think this should just be a logger.debug(...) and the logger's log level will switch this on or off.

Reply:
I don't want this to be debug level (I don't want all the other debugging logging when I have this on) and it is very verbose so I don't want it on all the time, so it deserves it's own flag.

Reply:
Ok.